ann_benchmarks/datasets.py

```diff
@@ -424,6 +424,60 @@ def lastfm(out_fn, n_dimensions, test_size=50000):
     # as the inner product on the untransformed data
     write_output(item_factors, user_factors, out_fn, 'angular')
 
+def movielens(file, out_fn, separator='::'):
+    print('preparing %s' % out_fn)
+
+    users = {}
+    X = []
+    dimension = 0
+    for line in file:
+        el = line.decode('UTF-8').split(separator)
+
+        userId = el[0]
+        itemId = int(el[1])
+        rating = float(el[2])
+
+        if rating < 3: # We only keep ratings >= 3
+            continue
+
+        if not userId in users:
+            users[userId] = len(users)
+            X.append([])
+
+        X[users[userId]].append(itemId)
+        dimension = max(dimension, itemId+1)
+
+    X_train, X_test = train_test_split(numpy.array(X), test_size=500, dimension=dimension)
+    write_sparse_output(X_train, X_test, out_fn, 'jaccard', dimension)
+
+def movielens1m(out_fn):
+    import zipfile
+
+    url = 'http://files.grouplens.org/datasets/movielens/ml-1m.zip'
+    fn = 'ml-1m.zip'
+    download(url, fn)
+    with zipfile.ZipFile(fn) as z:
+        movielens(z.open('ml-1m/ratings.dat'), out_fn)
+
+def movielens10m(out_fn):
+    import zipfile
+
+    url = 'http://files.grouplens.org/datasets/movielens/ml-10m.zip'
+    fn = 'ml-10m.zip'
+    download(url, fn)
+    with zipfile.ZipFile(fn) as z:
+        movielens(z.open('ml-10M100K/ratings.dat'), out_fn)
+
+def movielens20m(out_fn):
+    import zipfile
+
+    url = 'http://files.grouplens.org/datasets/movielens/ml-20m.zip'
+    fn = 'ml-20m.zip'
+    download(url, fn)
+    with zipfile.ZipFile(fn) as z:
+        f = z.open('ml-20m/ratings.csv')
+        f.readline()# ignore the header
+        movielens(f, out_fn, ',')
```

Comment:
Couldn't that be done by a single method by passing the file names as parameters? It looks somehow suspicious that ml-20m is different to 10m and 1m?

Reply:
Well I'm not sure of the why, but ml-20m is a csv, while 10m and 1m are .dat files, so the format is a bit different and I tried to merge their processing but it seems complicated to do it further (unless I pass an option to remove the headers of the csv file ?)

Reply:
Precisely, what about `def movielens(out_fn, fn, ignore_header=False)`?

Reply:
Since the zip files do not use the same hierarchy I added a parameter `ratings_file`, and for the different separator I added `separator`, giving:
```
movielens(fn, ratings_file, out_fn, separator='::', ignore_header=False)
```

would that be acceptable ?